apps/test_ogrsf.cpp

```diff
@@ -1144,6 +1178,85 @@ static int TestBasic( OGRLayer *poLayer )
         }
     }
 
+    // Test consistency of layer capabilities and driver metadata
+    if ( poDS )
+    {
+        GDALDriver *poDriver = poDS->GetDriver();
+
+        const bool bLayerShouldHaveEditCapabilities = !bReadOnly && !pszSQLStatement;
+
+        if ( poLayer->TestCapability(OLCMeasuredGeometries) && !poDriver->GetMetadataItem( GDAL_DCAP_MEASURED_GEOMETRIES ) )
+        {
+            bRet = FALSE;
+            printf("FAILURE: Layer advertizes OLCMeasuredGeometries capability but driver metadata does not advertize GDAL_DCAP_MEASURED_GEOMETRIES!\n" );
+        }
+        if ( poLayer->TestCapability(OLCCurveGeometries) && !poDriver->GetMetadataItem( GDAL_DCAP_CURVE_GEOMETRIES ) )
+        {
+            bRet = FALSE;
+            printf("FAILURE: Layer advertizes OLCCurveGeometries capability but driver metadata does not advertize GDAL_DCAP_CURVE_GEOMETRIES!\n" );
+        }
+        if ( poLayer->TestCapability(OLC25DGeometries) && !poDriver->GetMetadataItem( GDAL_DCAP_25D_GEOMETRIES ) )
+        {
+            bRet = FALSE;
+            printf("FAILURE: Layer advertizes OLC25DGeometries capability but driver metadata does not advertize GDAL_DCAP_25D_GEOMETRIES!\n" );
+        }
+
+        if ( poLayer->TestCapability(OLCCreateField) && !poDriver->GetMetadataItem( GDAL_DCAP_CREATE_FIELD ) )
+        {
+            bRet = FALSE;
+            printf("FAILURE: Layer advertizes OLCCreateField capability but driver metadata does not advertize GDAL_DCAP_CREATE_FIELD!\n" );
+        }
+        if ( bLayerShouldHaveEditCapabilities && poDriver->GetMetadataItem( GDAL_DCAP_CREATE_FIELD ) && !poLayer->TestCapability(OLCCreateField) )
+        {
+            bRet = FALSE;
+            printf("FAILURE: Driver metadata advertizes GDAL_DCAP_CREATE_FIELD but layer capability does not advertize OLCCreateField!\n" );
+        }
+
+        if ( poLayer->TestCapability(OLCCreateField) && !poDriver->GetMetadataItem( GDAL_DMD_CREATIONFIELDDATATYPES ) )
+        {
+            bRet = FALSE;
+            printf("FAILURE: Layer advertizes OLCCreateField capability but driver metadata does not include GDAL_DMD_CREATIONFIELDDATATYPES!\n" );
+        }
+        if ( bLayerShouldHaveEditCapabilities && poDriver->GetMetadataItem( GDAL_DMD_CREATIONFIELDDATATYPES ) && !poLayer->TestCapability(OLCCreateField) )
```

Comment:
the following would be safer
```suggestion
        if ( poDriver->GetMetadataItem( GDAL_DMD_CREATIONFIELDDATATYPES ) && !poDriver->GetMetadataItem( GDAL_DCAP_CREATE_FIELD ) )
```